### ISSUE TITLE
What can be done to achieve (system) symbol resolution for minidumps?

### ISSUE THREAD

**Issue by Rantanen** [NationalSecurityAgency/ghidra#2552]
I'm working on a Windows Minidump loader for Ghidra. The basic memory loading works fine currently, which is already a huge help in being able to position the modules in proper addresses to follow calls between modules.

The next thing I'd love to achieve is better symbol name resolution. I'm not completely familiar with the mechanism Ghidra uses for resolving symbol names, but I noted for PE files the system symbols are resolved even with no PDB files being available. The minidumps have module information embedded in them so I'm now wondering whether there is a way to communicate that to Ghidra to allow Ghidra to resolve the symbols?

I noted that some symbols were already resolved by the analyzer in some cases (`atoi`, `__CheckForDebuggerJustMyCode`), which made me hopeful that being able to resolve more of them wouldn't be too difficult.

If no one is able to answer this, my fallback plan will be to dig through the PE-loader and see how it provides the module information to Ghidra. Even if no one can describe the mechanism in full, I'm hoping someone could give at least a pointer at which parts of the PE loader I should start my research from.

I know analyzing Minidumps (ie. runtime state) is slightly off the usual use cases of reverse engineering programs, but my other option was tracking calls in Inkscape and that was getting unbearable. :)

**Comment by ryanmkurtz** [comment]
Does the Minidump file have the full PE header present?

**Comment by Rantanen** [comment]
Now that you mention it, it should. I'm not completely familiar with either one (the last few days have been a bit of a crash course into working with these files by hand), but as far as I know, the complete file image is part of the process memory and should be part of the dump.

The dump also lists the included modules, together with their base address where they are loaded in the memory: https://docs.microsoft.com/en-us/windows/win32/api/minidumpapiset/ns-minidumpapiset-minidump_module

The radare blog [has an article](https://radareorg.github.io/blog/posts/minidump/) that has extracting the PE file from the minidump as one of the examples. I did consider just doing that and then working with those files instead of trying to debug the dump as is - but that would lose the runtime information (process stack, etc.).

Re-using the PE information in loading the symbols sounds like a definite step towards the goal. I guess I'll just need to figure out how to feed the base address information to whatever part processes the PE stuff.

**Comment by ryanmkurtz** [comment]
Check out the `PortableExecutable` class.  It can be created with `SectionLayout.MEMORY` for your case, if you have an in-tact memory dump of the entire PE file.  You can look at the `PeLoader.load()` method to see how it uses the `PortableExecutable` class (with `SectionLayout.FILE` in this case).  You'll also see `PeLoader.load()` call `processSymbols()`, which may also help you.

**Comment by Rantanen** [comment]
Thanks! Those tips will definitely be useful when I'll get back to the project over the weekend!

**Comment by Rantanen** [comment]
Got my symbols loading! Thanks a lot for the pointers, they ended up working great!

In the end I ended up just calling into `PeLoader` directly (after copying the class over and exposing some of its internals as static methods). The big thing I couldn't directly reuse was the actual PE image loading.

- Instead of relying on `program.getBaseImage()`, which is set to the base address of the Minidump (in practice just `0x0`), I'm keeping track of the individual PE base addresses as defined in the Minidump `MINIDUMP_MODULE` struct and using that to calculate various section offsets.
- Include the name from `MINIDUMP_MODULE` in the sections (`HEADERS`, `.text´, etc.) to avoid duplicate names.
- Move the individual module fragments into their own folders to better organize them.

Once the individual `PortableExecutable` had been loaded and its memory mapped into the `Program` with section addresses resolved in a `Map`, the rest of the `processXyz` methods worked more or less as is. The only thing I had to comment out for now was the Ghidra comment handling, since those are relying on actual instance data and getting them working through static methods would have required moving the comment maps to parameters.

I do feel the code changes required to support using `PeLoader`/`AbstractPeDebugLoader` directly for this wouldn't be too bad if Ghidra ever wants to upstream this implementation, but for now I'll just keep using the copied loaders.

I'll close the issue since the original problem was answered and I feel a bit more comfortable digging through the Ghidra code base myself in case I encounter more similar problems.

With any luck I'll have something useful out next week.
